blockwork/workflows/workflow.py

```diff
@@ -12,94 +12,144 @@
-        'Recures elements and yields depths first'
-        for sub_element in element.iter_elements():
-            yield from self.depth_first_elements(sub_element)
-        yield element
-
-    # Note this is run as a post-init hook so subclasses can add attributes
-    # after the init by overriding the method but before the run.
-    @InitHooks.post
-    def run(self):
-        """
-        Run every transform from the configuration.
-
-        @ed.kotarski: This is a temporary implementation that I
-                      intend to change soon
-        """
-        # Join interfaces together and get transform dependencies and targets
+        return site_parser(SITE_TYPE).parse(site_path)
+
+    @classmethod
+    def parse_project(cls, ctx, site: Site, project: str, typ: type) -> Project:
+        'Parse a project option using the given site'
+        project_parser = parsers.Project(ctx, site)
+        project_path = Path(site.projects[project])
+        return project_parser(typ).parse(project_path)
+
+    @classmethod
+    def parse_target(cls, ctx, site: Site, project: Project, target:str, typ: type) -> base.Element:
+        'Parse a target option using the given site and project'
+        target_parser = parsers.Element(ctx, site, project)
+        # If target type not provided assume we're pointing to a workflow file itself
+        return target_parser.parse_target(target, typ)
+
+    def __call__(self, fn: Callable[..., Config]) -> Callable[..., None]:
+
+        @click.command()
+        @click.pass_obj
+        def command(ctx, project=None, target=None, *args, **kwargs):
+            site = self.parse_site(ctx)
+            if project:
+                kwargs['project'] = self.parse_project(ctx, site, project, self.project_type)
+                if target:
+                    kwargs['target'] = self.parse_target(ctx, site, kwargs['project'], target, self.target_type)
+
+            inst = fn(ctx, *args, **kwargs)
+            self._run(ctx, inst)
+
+        option_fns = []
+        if self.project_type:
+            option_fns.append(click.option('--project', '-p', type=str, required=True))
+        if self.target_type:
+            option_fns.append(click.option('--target', '-t', type=str, required=True))
+
+        # Apply the additional options
+        command = reduce(lambda f, o: o(f), option_fns, command)            
+
+        # This is a little horrible but this means click options
+        # can be added before or after the workflow decorator
+        command.params += getattr(fn, '__click_params__', [])
+
+        wf.add_command(command, name=self.name)
+        return command
+    
+
+    def gather(self, config: Config) -> Iterable[tuple[Config, list[Transform], list[Transform]]]:
+        '''
+        Iterate over the tree of configs gathering "interesting" transforms.
+        
+        yields [the config, its transforms, its target transforms]
```

Comment:
Make sure to use the Sphinx docstring format if you are providing commentary on what is returned, i.e. `:yields: ...`